pkg/oc/cli/cmd/debug.go

```diff
@@ -317,27 +306,80 @@ func (o *DebugOptions) Complete(cmd *cobra.Command, f kcmdutil.Factory, args []s
 		return err
 	}
 	o.ImageClient = imageClient.Image()
+
 	return nil
 }
+
 func (o DebugOptions) Validate() error {
+	if (o.AsRoot || o.AsNonRoot) && o.AsUser > 0 {
+		return fmt.Errorf("you may not specify --as-root and --as-user=%d at the same time", o.AsUser)
+	}
+	return nil
+}
+
+// Debug creates and runs a debugging pod.
+func (o *DebugOptions) RunDebug() error {
+	b := o.Builder().
+		WithScheme(ocscheme.ReadingInternalScheme, ocscheme.ReadingInternalScheme.PrioritizedVersionsAllGroups()...).
+		NamespaceParam(o.Namespace).DefaultNamespace().
+		SingleResourceType().
+		ResourceNames("pods", o.Resources...).
+		Flatten()
+	if len(o.FilenameOptions.Filenames) > 0 {
+		b.FilenameParam(o.ExplicitNamespace, &o.FilenameOptions)
+	}
+	one := false
+	infos, err := b.Do().IntoSingleItemImplied(&one).Infos()
+	if err != nil {
+		return err
+	}
+	if !one {
+		return fmt.Errorf("you must identify a resource with a pod template to debug")
+	}
+
+	templateV1, err := o.approximatePodTemplateForObject(infos[0].Object)
+	if err != nil && templateV1 == nil {
+		return fmt.Errorf("cannot debug %s: %v", infos[0].Name, err)
+	}
+	if err != nil {
+		glog.V(4).Infof("Unable to get exact template, but continuing with fallback: %v", err)
+	}
+	template := &kapi.PodTemplateSpec{}
+	if err := legacyscheme.Scheme.Convert(templateV1, template, nil); err != nil {
+		return err
+	}
+	pod := &kapi.Pod{
+		ObjectMeta: template.ObjectMeta,
+		Spec:       template.Spec,
+	}
+	pod.Name, pod.Namespace = fmt.Sprintf("%s-debug", generateapp.MakeSimpleName(infos[0].Name)), infos[0].Namespace
+	o.Attach.Pod = pod
+
+	if len(o.Attach.ContainerName) == 0 && len(pod.Spec.Containers) > 0 {
+		if len(o.FullCmdName) > 0 {
+			fmt.Fprintf(o.ErrOut, "Defaulting container name to %s.\n", pod.Spec.Containers[0].Name)
+			fmt.Fprintf(o.ErrOut, "Use '%s describe pod/%s -n %s' to see all of the containers in this pod.\n", o.FullCmdName, pod.Name, pod.Namespace)
+			fmt.Fprintf(o.ErrOut, "\n")
+		}
+
+		glog.V(4).Infof("Defaulting container name to %s", pod.Spec.Containers[0].Name)
+		o.Attach.ContainerName = pod.Spec.Containers[0].Name
+	}
+
 	names := containerNames(o.Attach.Pod)
 	if len(names) == 0 {
 		return fmt.Errorf("the provided pod must have at least one container")
 	}
-	if (o.AsRoot || o.AsNonRoot) && o.AsUser > 0 {
-		return fmt.Errorf("you may not specify --as-root and --as-user=%d at the same time", o.AsUser)
-	}
 	if len(o.Attach.ContainerName) == 0 {
```

Comment:
could these be moved to Validate as well?

Reply:
I didn't want to deal with the original flow so I left is as is, for now.